doc/kernel/usermode/syscalls.rst

```diff
@@ -0,0 +1,259 @@
+
+* Function pointers horribly confuse the limited parser. The workaround is
+  to typedef them first, and then express in the argument list in terms
+  of that typedef.
+
+* :c:macro:`__syscall` must be the first thing in the prototype and cannot
+  be preceded by anything else.
+
+Any header file that declares system calls must include a special generated
+header at the very bottom on the header file. This header follows the
+naming convention ``syscalls/<name of header file>``. For example, at the
+bottom of ``include/sensor.h``:
+
+.. code-block:: c
+
+    #include <syscalls/sensor.h>
+
+Implementation Details
+----------------------
+
+Declaring an API with :c:macro:`__syscall` causes some code to be generated in
+C and header files, all of which can be found in the project out directory
+under ``include/generated/``:
+
+* The system call is added to the enumerated type of system call IDs,
+  which is expressed in include/generated/syscall_list.h. It is the name
+  of the API in uppercase, prefixed with ``K_SYSCALL_``.
+
+* A prototype for the handler function is also created in
+  ``include/generated/syscall_list.h``
+
+* An entry for the system call is created in the dispatch table
+  ``_k_sycall_table``, expressed in ``include/generated/syscall_dispatch.c``
+
+* A weak handler function is declared, which is just an alias of the
+  'unimplemented system call' handler. This is necessary since the real
+  handler function may or may not be built depending on the kernel
+  configuration. For example, if a user thread makes a sensor subsystem
+  API call, but the sensor subsystem is not enabled, the weak handler
+  will be invoked instead.
+
+The body of the API is created in the generated system header. Using the
+example of :c:func:`k_sem_init()`, this API is declared in
+``include/kernel.h``. At the bottom of ``include/kernel.h`` is:
+
+::
+
+    #include <syscalls/kernel.h>
+
+Inside this header is the body of :c:func:``k_sem_init()``:
+
+::
+
+    K_SYSCALL_DECLARE3_VOID(K_SYSCALL_K_SEM_INIT, k_sem_init, struct k_sem *,
+                            sem, unsigned int, initial_count,
+                            unsigned int, limit);
+
+This generates an inline function which takes three arguments with void
+return value. Depending on context it will either directly call the
+implementation function or go through a system call elevation. A
+prototype for the implementation function is also automatically generated.
+
+The header containing :c:macro:`K_SYSCALL_DECLARE3_VOID()` is itself
+generated due to its repetitive nature and can be found in
+``include/generated/syscall_macros.h``.
+
+Implementation Function
+=======================
+
+The implementation function is what actually does the work for the API.
+Zephyr normally does little to no error checking of arguments, or does this
+kind of checking with assertions. When writing the implementation function,
+assume that all parameters have been validated, or check with assertions.
```

Comment:
This seems contradictory. Either you validate parameters or you don't.  Maybe just say validation of arguments (with assertions) is optional.